certstore_linux.go

```diff
@@ -1,14 +1,369 @@
+#include <p12plcy.h>
+#include <stdlib.h>
+#include <cryptohi.h>
 
-// This will hopefully give a compiler error that will hint at the fact that
-// this package isn't designed to work on Linux.
-func init() {
-	CERTSTORE_DOESNT_WORK_ON_LINIX
+SECItem *P12U_NicknameCollisionCallback(SECItem *old_nick, PRBool *cancel, void *wincx) {
+	char *nick = NULL;
+	SECItem *ret_nick = NULL;
+	CERTCertificate *cert = (CERTCertificate *)wincx;
+	if (!cancel || !cert) {
+		fprintf(stdout, "cancel: %p, cert: %p\n", cancel, cert);
+		return NULL;
+	}
+	nick = CERT_MakeCANickname(cert);
+	if (!nick) {
+		fprintf(stdout, "nick %p\n", nick);
+		return NULL;
+	}
+	if (old_nick && old_nick->data && old_nick->len &&
+		PORT_Strlen(nick) == old_nick->len &&
+		!PORT_Strncmp((char *)old_nick->data, nick, old_nick->len)) {
+		PORT_Free(nick);
+		fprintf(stdout, "old_nick %p, nick %p\n", old_nick, nick);
+		return NULL;
+    }
+	//fprintf(stdout, "using nickname: %s\n", nick);
+	ret_nick = PORT_ZNew(SECItem);
+	if (ret_nick == NULL) {
+		PORT_Free(nick);
+		fprintf(stdout, "ret_nick %p\n", ret_nick);
+		return NULL;
+	}
+	ret_nick->data = (unsigned char *)nick;
+	ret_nick->len = PORT_Strlen(nick);
+	return ret_nick;
+}
+*/
+import "C"
+import (
+	"crypto"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"syscall"
+	"unicode/utf16"
+	"unsafe"
+)
+
+type Passwd struct {
+	PwName   string
+	PwPasswd string
+	PwUid    int
+	PwGid    int
+	PwGecos  string
+	PwDir    string
+	PwShell  string
+}
+type nssStore int
+
+func (nssStore) Identities() ([]Identity, error) {
+	var identities = make([]Identity, 0)
+	//fmt.Printf("Listing certificates:\n")
+	var certs = C.PK11_ListCerts(C.PK11CertListType(C.PK11CertListAll), unsafe.Pointer(nil))
+	if certs == nil {
+		C.NSS_Shutdown()
+		return nil, errors.New(fmt.Sprintf("Error %d, closing and returing...\n", int(C.PR_GetError())))
+	}
+	var list *C.CERTCertList
+	var node *C.CERTCertListNode
+	list = certs
+	for node = CertListHead(list); ! CertListEnd(node, list); node = CertListNext(node) {
+		identities = append(identities, newNssIdentity(node))
+	}
+	return identities, nil
+}
+
+type nssIdentity struct {
+	node *C.CERTCertListNode
+}
+
+type nssPrivateKey struct {
+	node *C.CERTCertListNode
+}
+
+func (i *nssIdentity) Signer() (crypto.Signer, error) {
+	return i.newNssPrivateKey()
+}
+
+func (i *nssIdentity) Certificate() (*x509.Certificate, error) {
+	var der = i.node.cert.derCert
+	var bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
+	cert, err := x509.ParseCertificate(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	var der = i.node.cert.derCert
+	var bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
+	cert, err := x509.ParseCertificate(bytes)
+	if err != nil {
+		return nil, err
+	}
+	var certs = C.PK11_ListCerts(C.PK11CertListType(C.PK11CertListAll), unsafe.Pointer(nil))
```

Comment:
The compiler doesn't complain if I remove the explicit casts on Go 1.11

```suggestion
 	var certs = C.PK11_ListCerts(
 	C.PK11CertListAll, nil)
```

There are a few places where this could be cleaned up.

Reply:
will check i got 1.10